### PR TITLE
Fix test that leaks images.

### PR DESCRIPTION
**Pull request by polina-c**
Contributes to https://github.com/flutter/flutter/issues/145599

Repro:
`flutter test test/cupertino/tab_scaffold_test.dart  --dart-define LEAK_TRACKING=true --plain-name "Adding new tabs does not crash the app"`

What is going on: 
1. ImageCache.putIfAbsent, in case the image already existed in the cache, invokes ImageCache._trackLiveImage, that creates _LiveImage and passes the image's completer as parameter `completer`
3. _LiveImage constructor invokes super constructor (of _CachedImageBase) that initializes the member `handle`
4. `handle` is disposed [using scheduler](https://github.com/polina-c/flutter/blob/c698e694c871278fd00dc52394a91323494b6395/packages/flutter/lib/src/painting/image_cache.dart#L633), and disposal does not happen in time of test completion.

Adding delay to the test increases number of not disposed objects from 30 to 120.
Should we force schedule at the end of the widget tests somehow to make scheduler switched to right state?

Creation call stack:

```
#9______new_ImageStreamCompleterHandle.__(package:flutter/src/painting/image_stream.dart:465:41)
#10_____ImageStreamCompleter.keepAlive_(package:flutter/src/painting/image_stream.dart:655:39)
#11_____new__CachedImageBase_(package:flutter/src/painting/image_cache.dart:609:27)
#12_____new__LiveImage_(package:flutter/src/painting/image_cache.dart:647:9)
#13_____ImageCache._trackLiveImage.<anonymous_closure>_(package:flutter/src/painting/image_cache.dart:302:14)
#14______LinkedHashMapMixin.putIfAbsent_(dart:collection-patch/compact_hash.dart:543:23)
#15_____ImageCache._trackLiveImage_(package:flutter/src/painting/image_cache.dart:296:17)
#16_____ImageCache.putIfAbsent_(package:flutter/src/painting/image_cache.dart:378:7)
#17_____ImageProvider.resolveStreamForKey_(package:flutter/src/painting/image_provider.dart:517:81)
                      #18_____ScrollAwareImageProvider.resolveStreamForKey_(package:flutter/src/widgets/scroll_aware_image_provider.dart:104:19)
#19_____ImageProvider.resolve.<anonymous_closure>_(package:flutter/src/painting/image_provider.dart:366:9)
#20_____ImageProvider._createErrorHandlerAndKey.<anonymous_closure>_(package:flutter/src/painting/image_provider.dart:479:24)
#21_____SynchronousFuture.then_(package:flutter/src/foundation/synchronous_future.dart:43:39)
                      #22_____ImageProvider._createErrorHandlerAndKey_(package:flutter/src/painting/image_provider.dart:476:9)
#23_____ImageProvider.resolve_(package:flutter/src/painting/image_provider.dart:363:5)
#24______ImageState._resolveImage_(package:flutter/src/widgets/image.dart:1111:16)
#25______ImageState.didChangeDependencies_(package:flutter/src/widgets/image.dart:1061:5)
#26_____StatefulElement._firstBuild_(package:flutter/src/widgets/framework.dart:5630:11)
#27_____ComponentElement.mount_(package:flutter/src/widgets/framework.dart:5457:5)
#28_____Element.inflateWidget_(package:flutter/src/widgets/framework.dart:4334:16)
#29_____Element.updateChild_(package:flutter/src/widgets/framework.dart:3843:18)
#30_____SingleChildRenderObjectElement.mount_(package:flutter/src/widgets/framework.dart:6763:14)
```

### CODE REVIEW

**Comment by polina-c** [comment]
Creation call stack:

```
#9______new_ImageStreamCompleterHandle.__(package:flutter/src/painting/image_stream.dart:465:41)
                      #10_____ImageStreamCompleter.keepAlive_(package:flutter/src/painting/image_stream.dart:655:39)
                      #11_____new__CachedImageBase_(package:flutter/src/painting/image_cache.dart:609:27)
                      #12_____new__LiveImage_(package:flutter/src/painting/image_cache.dart:647:9)
                      #13_____ImageCache._trackLiveImage.<anonymous_closure>_(package:flutter/src/painting/image_cache.dart:302:14)
                      #14______LinkedHashMapMixin.putIfAbsent_(dart:collection-patch/compact_hash.dart:543:23)
                      #15_____ImageCache._trackLiveImage_(package:flutter/src/painting/image_cache.dart:296:17)
                      #16_____ImageCache.putIfAbsent_(package:flutter/src/painting/image_cache.dart:378:7)
                      #17_____ImageProvider.resolveStreamForKey_(package:flutter/src/painting/image_provider.dart:517:81)
                      #18_____ScrollAwareImageProvider.resolveStreamForKey_(package:flutter/src/widgets/scroll_aware_image_provider.dart:104:19)
                      #19_____ImageProvider.resolve.<anonymous_closure>_(package:flutter/src/painting/image_provider.dart:366:9)
                      #20_____ImageProvider._createErrorHandlerAndKey.<anonymous_closure>_(package:flutter/src/painting/image_provider.dart:479:24)
                      #21_____SynchronousFuture.then_(package:flutter/src/foundation/synchronous_future.dart:43:39)
                      #22_____ImageProvider._createErrorHandlerAndKey_(package:flutter/src/painting/image_provider.dart:476:9)
                      #23_____ImageProvider.resolve_(package:flutter/src/painting/image_provider.dart:363:5)
                      #24______ImageState._resolveImage_(package:flutter/src/widgets/image.dart:1111:16)
                      #25______ImageState.didChangeDependencies_(package:flutter/src/widgets/image.dart:1061:5)
                      #26_____StatefulElement._firstBuild_(package:flutter/src/widgets/framework.dart:5630:11)
                      #27_____ComponentElement.mount_(package:flutter/src/widgets/framework.dart:5457:5)
                      #28_____Element.inflateWidget_(package:flutter/src/widgets/framework.dart:4334:16)
                      #29_____Element.updateChild_(package:flutter/src/widgets/framework.dart:3843:18)
                      #30_____SingleChildRenderObjectElement.mount_(package:flutter/src/widgets/framework.dart:6763:14)
```

File: packages/flutter/test/cupertino/tab_scaffold_test.dart [hunk 1]
```diff
@@ -685,6 +688,8 @@ void main() {
 
     // Tapping the tabs should still work.
     expect(tabsPainted, const <int>[0, 0, 18]);
+
+    await tester.runAsync(() async =>Future<void>.delayed(const Duration(milliseconds: 10)));
```
Comment (Hixie) [inline packages/flutter/test/cupertino/tab_scaffold_test.dart:692]: replacing this line with `imageCache.clear();` fixes all the leaks in this test for me.

**Comment by polina-c** [comment]
Hm, on my mac there are less leaks with this change and they are different:

```
  Expected: leak free
    Actual: <Instance of 'Leaks'>
     Which: contains leaks:
            # The text is generated by leak_tracker.
            # For leak troubleshooting tips open:
            # https://github.com/dart-lang/leak_tracker/blob/main/doc/TROUBLESHOOT.md
            notDisposed:
              total: 6
              objects:
                StatefulElement:
                  test: Adding new tabs does not crash the app
                  identityHashCode: 823867307
                _MediaQueryFromViewState:
                  test: Adding new tabs does not crash the app
                  identityHashCode: 701933900
                LeafRenderObjectElement:
                  test: Adding new tabs does not crash the app
                  identityHashCode: 637995193
                RenderErrorBox:
                  test: Adding new tabs does not crash the app
                  identityHashCode: 761576866
                PictureLayer:
                  test: Adding new tabs does not crash the app
                  identityHashCode: 1042115368
                _NativePicture:
                  test: Adding new tabs does not crash the app
                  identityHashCode: 515013832
```
But bots run on Windows, and if on Windows there are no leaks, it is good fix for now. Eventually we will add bots for other platforms

However there is a question.
There are many tests that deal with animation. Why they are not leaking? Why we need to clean cash for this specific test. I am puzzled.

**Comment by Hixie** [comment]
This test loads images. The images get cached. My theory about it being related to animated GIFs was mistaken.

**Comment by Hixie** [comment]
We should document somewhere that if you use cached images, they'll get flagged as leaked unless you manually clear the cache.

File: packages/flutter/test/cupertino/tab_scaffold_test.dart [hunk 2]
```diff
@@ -47,6 +47,10 @@ void main() {
     selectedTabs = <int>[];
   });
 
+  tearDownAll(() {
+    imageCache.clear();
+  });
```
Comment (Hixie) [inline packages/flutter/test/cupertino/tab_scaffold_test.dart:52]: Per https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#make-each-test-entirely-self-contained it's generally better to put this in each test, that way you can easily copy tests to other files and generally easily tell what exactly each test is doing.

**Comment by Hixie** [comment]
cc @jonahwilliams who is probably the person most familiar with the image cache?
cc @goderbauer and @Piinks who may have opinions too

File: packages/flutter/test/cupertino/tab_scaffold_test.dart (hunk 2 above)
Comment (polina-c) [reply packages/flutter/test/cupertino/tab_scaffold_test.dart:52]: Clean up for each test will improve test isolation.
Clean up just at the end will improve test performance.

I guess one of the reasons image cashing exists is exactly performance of testing.

But I am open for both ways.

**Comment by polina-c** [comment]
> We should document somewhere that if you use cached images, they'll get flagged as leaked unless you manually clear the cache.

Added the guidance to the page that is linked in leak related exceptions: https://github.com/dart-lang/leak_tracker/pull/233

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**